backend/example_bib_app/services.py

```diff
@@ -0,0 +1,135 @@
+
+The `AuthorService` class defines `filterset_fields` instead of `filterset_class`. This attribute allows for filtering of the queryset based on exact matches.
+ For more complex filtering, the `filterset_class` attribute can be used instead.
+
+The `PublisherService` class defines `filterset_class` instead of `filterset_fields`. This attribute allows for filtering of the queryset using a custom filter class named `PublisherFilterSet`.
+
+The `BookService` class also defines `filterset_class` using a custom filter class named `BookFilterSet`. Additionally, the `BookStreamingService` class is defined to stream books to the client. 
+The `Stream` method is decorated with the `@grpc_action` decorator, which specifies the request and response message types, as well as the request and response stream types.
+
+Overall, this module provides a convenient way to define gRPC services for the specified models in the `example_bib_app` application. To improve the readability of the code, 
+the `filterset_fields` and `filterset_class` attributes could be grouped by their type (e.g., author-related filters, publisher-related filters) to make it easier to see which filters are being used. 
+Additionally, the `BookStreamingService` class could be moved to a separate module to improve the organization of the code.
+"""
+from django_socio_grpc import generics
+from django_socio_grpc.decorators import grpc_action
+
+from .models import Author, Publisher, PublicationCategory, Book, Journal
+
+from .serializers import AuthorProtoSerializer, PublisherProtoSerializer, PublicationCategoryProtoSerializer, BookProtoSerializer, JournalProtoSerializer
+
+from .filters import PublisherFilterSet, BookFilterSet
+
+class AuthorService(generics.AsyncModelService):
+    queryset = Author.objects.all()
+
+    # filterset_fields allows for filtering of the queryset, only based on exact matches.
+    # for more complex filtering, use filterset_class
+    filterset_fields = ('name_first', 'name_last',)
+    serializer_class = AuthorProtoSerializer
+
+class PublisherService(generics.AsyncModelService):
+    queryset = Publisher.objects.all()
+
+    # the optional filterset_class allows for filtering of the queryset (s. filter client example)
+    filterset_class = PublisherFilterSet
+    serializer_class = PublisherProtoSerializer
+
+class PublicationCategoryService(generics.AsyncModelService):
+    queryset = PublicationCategory.objects.all()
+    serializer_class = PublicationCategoryProtoSerializer
+
+class BookService(generics.AsyncModelService):
+    queryset = Book.objects.all()
+    filterset_class = BookFilterSet
+    serializer_class = BookProtoSerializer
+
+    @grpc_action(
+        response=[
+            {
+                "name": "book",
+                "type": "BookResponse",
+            }
+        ],
+        response_stream=True,
+    )
+    async def StreamAllBooks(self, request, context):
```

Comment:
note that there is already a StreamMixin in the django-socio-grpc framework. So please make a specific behavior of the stream and use the default one by DSG to show user that it exist: https://github.com/socotecio/django-socio-grpc/blob/master/django_socio_grpc/mixins.py#L399

Reply:
Thanks, @AMontagu,
for the hint - I was not aware of this.
I added it in the latest commit. But there is an Error, when I call the Stream RPC:

``grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "'InternalHttpRequest' object has no attribute 'write'"
        debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2023-11-23T22:19:57.956831673+01:00", grpc_status:2, grpc_message:"\'InternalHttpRequest\' object has no attribute \'write\'"}"``

Could you please check, why this is happening ?
See, e.g. bib_example_client.py 

Merci

Reply:
I will look into it. This is possible the read/write api not working since we improved our internal system